### PR TITLE
Pseudo experiment analysis

### PR DESCRIPTION
**Pull request by ajlee21**
This PR adds the final set of analyses that will go into the main text of the manuscript

This analysis is generating pseudo-experiments using the VAE and looking at the 1) consistency of the pseudo-experiment vs the real experiment and 2) comparing the representation of the expression data using pseudo-experiments created using the experiment-level vs sample-level simulation.

The main scripts to focus on are:
DE_analysis_run.R
find_enrichment_run.R (find_enrichment_individual.R is similar)
create_heatmap.ipynb
generate_random_sampled_experiment.ipynb
generate_E_GEOD_51409_template_experiment.ipynb (generate_additional_template_experiments.ipynb is very similar)

### CODE REVIEW

File: Pseudo_experiments/create_heatmap.ipynb [hunk 1]
```diff
@@ -0,0 +1,1141 @@
...
...58+fh4+MDKysrHD58GCtXrqxyT+aD2AwSERER/UXJyckYPHgwXF1djcb79euHuLg4bN68GaNGjcKMGTOwefNmw3y7du3g4uKCnTt3AgDmzp2LBQsWoFevXigpKUHTpk0xc+ZM+Pn5Gd5TUlKCLVu2YMqUKZVy3LlzB/Pnz0dubi5UKhWaNWuGN954A8OGDat2G1SSnP2H9LeLjY195ItB/sp7qhPadtrfur5aSaUSncAkqEorREcwCdlvNBQdwSRI/LGTRbLkr3A5zkZPFB1B0XifQSIiIiIzxmaQiIiIyIyxGSQiIiIyY2wGiYiIiMwYm0EiIiIiM8ZmkIiIiMiMsRkkIiIiMmNsBomIiIjMGJ9AIoiNjc0j30D67Nmzf3sO6bdzf/s6ax1JLzqBSdCVlYmOYBLUtxqLjmASLEtEJzANEnfp0N+AzaAgkydPfuT3/N1PHyEiIiLi3xREREREZozNIBEREZEZYzNIREREZMbYDBIRERGZMTaDRERERGaMzSARERGRGWMzSERERGTG2AwSERERmTE2g0RERERmjM0gERERkRljM0hERERUjeHDh8Pb2xs+Pj7w9fVFeHg4tm3bZrTMtGnT4O7ujoMHD1Z6/40bNxATEwMfHx906tQJCxcuhF6vN8xfu3YNb775Jjp16gQ/Pz+88sorOHnypGFep9Ph008/RUhICHx8fDB06FCj+S1btsDHx8foy9PTE1FRUdVuG5tBIiIiIhmio6Nx5MgRZGVl4YUXXsD48eORk5MDANBqtUhPT4ejoyMSExMrvXfixIkAgN27d2Pjxo3YsWMHvvjiC8P8rFmzcOvWLWRmZmLv3r3w9vbG66+/DkmSAAArVqzAli1bsGrVKmRlZaFjx46IiIiAVqsFAISFheHIkSOGr927d8Pa2hphYWHVbhebQSIiIqJHYGVlhaFDh0Kn0+HUqVMA7u6Z02g0mD59OrZv346CggLD8ufPn8e+ffswadIkODg4wNXVFREREUZNY25uLkJDQ+Ho6AiNRoOBAwfi8uXLhvVkZmZi6NChcHV1hUajwdixY3Hz5k3s2LGjyoypqamws7NDjx49qt0eNoNERERUqxUUFCAnJ6fS1/0N26MoKyvDunXroFar4eHhAQBITk5Gnz59EBoaCjs7O6SkpBiWz87OhoODA1q0aGEY8/LywsWLFw179kaNGoX
```
Comment (arielah) [inline Pseudo_experiments/create_heatmap.ipynb]: This heatmap looks really good! Me being a naive reader, I don't know how to interpret the heatmap, so I might appreciate a sentence or two of explanation about what I'm looking at within the notebook. But if the heatmaps are being made for a manuscript, then that explanation's probably in the text, so your call.

File: Pseudo_experiments/generate_E_GEOD_51409_template_experiment.ipynb [hunk 2]
```diff
@@ -0,0 +1,2057 @@
...
+    "import ast\n",
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "import seaborn as sns\n",
+    "import random\n",
+    "import glob\n",
+    "from sklearn import preprocessing\n",
+    "\n",
+    "sys.path.append(\"../\")\n",
+    "from functions import utils\n",
```
Comment (arielah) [inline Pseudo_experiments/generate_E_GEOD_51409_template_experiment.ipynb]: Not relevant to anything, I just found this line funny

File: Pseudo_experiments/nbconverted/generate_E_GEOD_51409_template_experiment.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                                columns=original_data.columns,
+                                index=original_data.index)
+
+original_data_scaled_df.head(5)
+
+
+# In[12]:
+
+
+# Re-scale simulated data back into the same range as the original data
```
Comment (arielah) [inline Pseudo_experiments/nbconverted/generate_E_GEOD_51409_template_experiment.py]: Just confused, why are you scaling the original data and then un-scaling the simulated data? I got a little turned around in this part.

File: Pseudo_experiments/generate_random_sampled_experiment.ipynb [hunk 4]
```diff
@@ -0,0 +1,1453 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Explore randomly-sampled experiments\n",
+    "\n",
+    "This notebook examines experiments generated by the random sampling approach (sample level simulation).  We use this as a control for our pseudo experiments generated the experiment-preserving approach that is implemented in ```generate_E_GEOD_51409_template_experiment.ipynb```"
```
Comment (arielah) [inline Pseudo_experiments/generate_random_sampled_experiment.ipynb]: Took me a couple of reads to understand this comment, but this script is you developing the negative control dataset, with generate_E_GEOD_51409_template_experiment is what it's compared to?

**Comment by arielah** [comment]
I can't remember what I meant to say that got lost, and I'm certain it's not important, so this should be all my feedback. Good job!

File: Pseudo_experiments/DE_analysis_run.R [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+library("limma")
+
+get_DE_stats <- function(metadata_file, 
+                      experiment_id, 
+                      expression_file,
+                      data_type,
+                      run){
+  
+  # Read in data
+  expression_data <- t(as.matrix(read.table(expression_file, sep="\t", header=TRUE, row.names=1)))
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply Pseudo_experiments/DE_analysis_run.R]: Good question. All my scripts deal with expression data as a matrix of sample x gene, while lmfit requires the input as gene x sample.  I'll add a comment for my future self/users just so its clear

File: Pseudo_experiments/DE_analysis_run.R [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  metadata <- as.matrix(read.table(metadata_file, sep="\t", header=TRUE, row.names=1))
+  
+  # NOTE: It is very,very,very important that you make sure the metadata is in the same order
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply Pseudo_experiments/DE_analysis_run.R]: This was a quick-ish one off analysis so I did more manual checks this time. But definitely wouldn't mind incorporating this feature if you want to share your scripts

File: Pseudo_experiments/DE_analysis_run.R [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  
+  ## DEGs of simulated data
+  # lmFit fits a linear model using weighted least squares for each gene:
+  fit <- lmFit(expression_data, mm)
+  
+  # Comparisons between groups (log fold-changes) are obtained as contrasts of these fitted linear models:
+  # Specify which groups to compare:
+  
+  if (grep("metadata_deg_temp", metadata_file)){
+    contr <- makeContrasts(group37 - group22, levels = colnames(coef(fit)))
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply Pseudo_experiments/DE_analysis_run.R]: These grouping depend on the experiment that is being used. In this case I was measuring the differential gene expression using an experiment with two different growth conditions (samples grown in 22 degrees versus samples grown in 37 degrees). I'll add a comment for this though

File: Pseudo_experiments/DE_analysis_run.R [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  all_genes <-  as.data.frame(top.table)
+  
+  # Find all DEGs based on adjusted p-value cutoff
+  # These p-values are FDR adjusted
+  sign_DEGs <- all_genes[all_genes[,'adj.P.Val']<0.05,]
+  
+  # Save summary statistics of DEGs
+  if (data_type == "control"){
+    #out_sim_filename = paste("/home/alexandra/Documents/Data/Batch_effects/pseudo_experiment/output_control/DEG_control_data_E-GEOD-51409_", run, ".txt", sep="")
+    out_sim_filename = paste("~/UPenn/CGreene/Remote/pathway_analysis/output_control/DE_stats_control_data_E-GEOD-51409_", run, ".txt", sep="")
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply Pseudo_experiments/DE_analysis_run.R]: Good point! I'm not sure how the working directories work in R, but I thought I could just call ```filename='filename.txt'``` if the file was in the same directory as my R script. But it couldn't find the file without me specifying the full path.

File: Pseudo_experiments/find_enrichment_individual.R [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  # Convert pathway data into proper format, which looks to be a dictionary
+  # Pathway name : list of genes in the pathway
+  # Note: parsing is currently based on the format obtained from annotation files 
+  # downloaded from ADAGE
+  pathway_parsed <- {}
+  for (i in 1:nrow(pathway)){
+    pathway_parsed[i] <-pathway_names[i]
+  }
+  for (i in 1:nrow(pathway)){
+    pathway_parsed[paste(pathway_names[i])] <- strsplit(pathway[i,2],";")
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply Pseudo_experiments/find_enrichment_individual.R]: Ok I wasn't sure if a dictionary equivalent did exist in R so felt silly for doing this. It looked like the function needed to have the inputs in this format, but perhaps its some other R type that I didn't understand. I'm not sure if you're familiar with this fgsea library?

File: Pseudo_experiments/create_heatmap.ipynb (hunk 1 above)
Comment (ajlee21) [reply Pseudo_experiments/create_heatmap.ipynb]: Very good suggestion! I will add this

File: Pseudo_experiments/nbconverted/generate_E_GEOD_51409_template_experiment.py (hunk 3 above)
Comment (ajlee21) [reply Pseudo_experiments/nbconverted/generate_E_GEOD_51409_template_experiment.py]: This is definitely confusing. I will add a comment about this. 

Originally the expression data was 0-1 normalized for use in training the VAE, however when we started performing followup DE analyses and pathway analyses we found that the normalized data had reduced variance that resulted in an inconsistency between the number of DEGs found compared to the publication. So we are re-scaling/un-scaling our normalized data to be in the original range of data.

File: Pseudo_experiments/generate_random_sampled_experiment.ipynb (hunk 4 above)
Comment (ajlee21) [reply Pseudo_experiments/generate_random_sampled_experiment.ipynb]: You're correct! I'm comparing simulated experiments that were simulated differently. I have updated the comment in the notebook so hopefully is makes more sense.

**Comment by ajlee21** [comment]
I have updated the scripts based on your very helpful comments. Let me know if there are any additional comments or question you have.

I will also be updating the scripts based on feedback from my 1-1 meeting

File: Pseudo_experiments/nbconverted/create_heatmap.py [hunk 10]
```diff
@@ -286,11 +286,19 @@
...
 
 
 # Plot control
 selected_control_DEG_data = selected_control_data[control_gene_ids]
 f = sns.clustermap(selected_control_DEG_data.T, cmap="viridis")
 f.savefig(heatmap_control_file, dpi=500)
 
+
+# **Summary**
+# The heatmaps display the top differentially expressed genes (FDR adj p-value < 0.05 and log2 FC > 1) identified using the original expression data (```selected_original_DEG_data```), using simulated expression data created by randomly sampling the compendium (```selected_control_DEG_data```), using the simulated expression data created by sampling by experiment from compendium (```selected_simulated_DEG_data```).
```
Comment (arielah) [inline Pseudo_experiments/nbconverted/create_heatmap.py]: Really helpful explanation!

File: Pseudo_experiments/nbconverted/generate_E_GEOD_51409_template_experiment.py [hunk 11]
```diff
@@ -130,6 +133,8 @@
 
 
 # ## Process data
+# 
+# Notice: Originally the expression data was 0-1 normalized for use in training the VAE, however when we performed differential expression analyses we found that the normalized data had reduced variance that resulted in an inconsistency between the number of DEGs found compared to the publication. Therefore, we are re-scaling our normalized data to be in the original range of data.
```
Comment (arielah) [inline Pseudo_experiments/nbconverted/generate_E_GEOD_51409_template_experiment.py]: Just a curiosity thing: is this repo going to be public when the manuscript is published? I'm wondering if your code will have to be understandable to random passersby like me or if the documentation is just for future you and other Greene lab members who might want to continue the project. Either way, I think you're in good shape.

Comment (ajlee21) [reply Pseudo_experiments/nbconverted/generate_E_GEOD_51409_template_experiment.py]: This repo is already public, but will be highlighted in the manuscript that we're working on. I anticipate outside users may want to use the simulation notebooks, but not these type of followup experiments that more to support using the VAE to simulate data.  So I'm not as concerned about these scripts in that sense.

**Pull request merged**